src/expr/parser.rs

```diff
@@ -8,53 +8,76 @@ use nom::digit;
 use std::str;
 use std::str::FromStr;
 
-named!(parens<i64>, delimited!(
+use expr::expr::Expr;
+use expr::expr::BinOp;
+
+use std::mem;
+
+named!(parens<Expr>, delimited!(
     char!('('),
     expr,
     char!(')')
   )
 );
 
-named!(i64_literal<i64>,
-  map_res!(
-    map_res!(
-      digit,
-      str::from_utf8
-    ),
-    FromStr::from_str
-  )
+named!(i64_literal<Expr>,
+       map_res!(
+           map_res!(
+               digit,
+               str::from_utf8),
+           Expr::from_str)
 );
 
-named!(factor<i64>,
+named!(factor<Expr>,
   alt!(
     i64_literal
   | parens
   )
 );
 
 // we define acc as mutable to update its value whenever a new term is found
-named!(term <i64>,
-  chain!(
-    mut acc: factor  ~
-             many0!(
-               alt!(
-                 tap!(mul: preceded!(tag!("*"), factor) => acc = acc * mul) |
-                 tap!(div: preceded!(tag!("/"), factor) => acc = acc / div)
-               )
-             ),
-    || { return acc }
-  )
-);
+named!(term <Expr>,
+       chain!(
+           mut acc: factor  ~
+               many0!(
+                   alt!(
+                       map!(preceded!(tag!("*"), factor),
+                            |mul| {
+                                let old = mem::replace(&mut acc, Expr::Literal(0));
+                                mem::replace(&mut acc, Expr::BinOp(Box::new(old), BinOp::Times, Box::new(mul)));
+                                ()
```

Comment:
I believe this closing `()` is unnecessary. Ending with a `;` should be equivalent.